arxiv/integration/api/service.py

```diff
@@ -0,0 +1,261 @@
+"""
+Provides a base class for HTTP API integrations.
+
+The goal of this module is to handle boilerplate connection and request
+handling that is invariant (or nearly so) among service integration modules
+that target HTTP APIs.
+
+Specific goals
+--------------
+
+- Handle verifying the response status code, and raise appropriate and
+  semantically meaningful exceptions.
+- Handle setting up a persistent HTTP session to cut down on overhead.
+- Provide basic retry functionality.
+- Handle binding a service instance (with its persistent session) to the
+  Flask application context.
+
+.. todo::
+   Make retry parameters easier to configure.
+
+"""
+
+from typing import Optional, Tuple, MutableMapping, List
+import inspect
+import requests
+from urllib3.util.retry import Retry
+from urllib.parse import urlparse, urlunparse, urlencode
+import json
+
+from flask import g, Flask, current_app
+
+from . import status
+from ..meta import MetaIntegration
+import logging
+
+from .exceptions import RequestFailed, RequestUnauthorized, RequestForbidden, \
+    NotFound, BadRequest, SecurityException, ConnectionFailed, BadResponse
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_ENDPOINT = 'http://fooservice:8000/'
+DEFAULT_VERIFY = True
+
+
+class HTTPIntegration(metaclass=MetaIntegration):
+    """
+    Integrate with a service over HTTP.
+
+    Provides binding to the application context, so that connections can be
+    re-used.
+
+    To implement an HTTP API service integration, define a class that inherits
+    from :class:`.HTTPIntegration`. At minimum, it should include an inner
+    class called ``Meta`` with the attribute ``service_name``; this is used to
+    identify the configuration parameters for the service. For example:
+
+    .. code-block:: python
```

Comment:
The COOL_BEANS_ENDPOINT mapping from a config to a class/instance seems circuitous and unexpected.

It seems like this fixes the implementation of the service at each place that imports and uses it. 

What is the advantage of this over how jinja adds a dict entry to the app? Such as in factory.py create_web_app() doing something like app['CoolIntegration'] = CoolIntegrationImplXyz( ... )

Reply:
Great question; will reply to this and https://github.com/arXiv/arxiv-base/pull/107#discussion_r266102345 here.

Last thing first: this implementation totally allows for direct instantiation, e.g. to use the same implementation with different configs in the same app. This works as expected:

```python
fm = FileManagementServiceImplXYZ( config.get(xxx)...)
fm2 = FileManagementServiceImplXYZ( config.get(yyy)...)
fm.get_upload(id)
fm2.get_upload(id)
```

One of the goals here, that could probably use some elaboration in [this slightly stale documentation](https://arxiv.github.io/arxiv-arxitecture/crosscutting/services.html#controller-modules) is to not be importing and using Flask proxy objects (most notably ``current_app``) all over the place. In particular, a requirement for controller modules in arXiv NG apps is that they avoid use of ``current_app`` and other proxy objects, in order to cut down on side-effects and reduce the amount of mocking required to test them. That's part of why we are separating "routes" from "controllers" -- Flask integration (setting up routing, working with the ``request`` proxy, etc) gets concentrated in the routes module(s), leaving the controllers to focus on the logic of handling the request.

What that means for service integration modules is that they should be handling Flask context management themselves. This mirrors the Flask extension pattern. The goal is to be able to simply import and use the integration module in controllers, and not worry about Flask concerns like contexts.

So, specifically to the question 
> What is the advantage of this over how jinja adds a dict entry to the app?

Doing this in your controllers would not be consistent with those goals:

```python
from flask import current_app

def handle_some_request(params, etc):
    current_app.someService.do_something()

```

Indeed, that's not very helpful if we are truly using the same implementation to talk to more than one service with the same API in the same app. I'd be interested to hear if there are cases where we need to do this.

Reply:
I'm for the goal of removing the flask current_app and request imports.

A simpler the way to achieve these goals:

1. app.py create_web_app() sets up app with services from config or env vars or somewhere else.

2. In routes.py the blueprint code provides controllers services. The blueprint code will get called to handle the route, it will then get the service from the app dict, then it will call the controller with it needed services as arguments.

This satisfies the goals of not importing flask proxy objects and makes mocking trivial.

It also satisfies the goal of not binding the service implementation class in the controller.

This is also a pattern of use that all developers will be familiar with and well supported by development tools.

This pattern has the benefit of no code to maintain. 

route.py
```python
@blueprint.route('abs/<path:arxiv_id>', methods=['GET'])
def abstract(arxiv_id: str) -> Response:
    """Abstract (abs) page view."""

    abs_doc_service = current_app.abs_doc_service
    response, code, headers = abs_page.get_abs_page(abs_doc_service, arxiv_id)

    if code == status.HTTP_200_OK:
        return render_template('abs/abs.html', **response), code, headers
```

The blueprint is responsible to provide the abs_page controller function access to any services it needs.  The blueprint/route seems like a good place to do it because it is already deep in the territory of flask proxy imports.

Reply:
Yeah, dependency injection might be a reasonable way to approach this. Since this would be a departure from how I've asked the team to implement controllers, routes, and services heretofore, we'll need to be deliberate and systematic about making such a change.

If you'd like to put together a brief (one-page) summary outlining the proposed change, we can get input from the team, which will help to inform a final decision. For current on-deck releases (submission-0.1, browse-0.2, probably browse-0.3) we'll proceed as I've outlined above.

Reply:
I'm proposing something more along the lines of  "threading the state" than capital D capital I Dependency Injection.

Reply:
Sure, fair enough